### PR TITLE
Integrate DfE::Analytics

### PR DESCRIPTION
**Pull request by duncanjbrown**
## Context

We now have https://github.com/DFE-Digital/dfe-analytics to replace custom code in all the apps.

## Changes proposed in this pull request

Delete the old BigQuery code. Add DfE::Analytics. Retain the old worker so we don't fall over on jobs enqueued during the change.

## Guidance to review

Use the `log_only` and `async` options in `dfe-analytics.rb` to satisfy yourself everything's working. DM me for BigQuery credentials if you need them.

## Link to Trello card

https://trello.com/c/neK2rSqu/563-share-bigquery-client-code-across-bat-services-dfe-analytics

## Things to check

- [X] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [X] This code does not rely on migrations in the same Pull Request
- [X] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [X] API release notes have been updated if necessary
- [X] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Comment by duncanjbrown** [comment]
Validation error tracking depends on `ApplicationHelper.service_key` which somehow isn't in scope any more. I'm having to explicitly include `ApplicationHelper` in `ApplicationController` for some reason, which seems reasonable, but I'm now not sure why this worked in the first place. Investigating

**Comment by duncanjbrown** [comment]
It was included via `EmitRequestEvents`. This was obviously wrong. Fixed

File: config/initializers/dfe-analytics.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  # Passed directly to the timeout: option on the BigQuery client
+  #
+  # config.bigquery_timeout = 120
+
+  # A proc which returns true or false depending on whether you want to
+  # enable analytics. You might want to hook this up to a feature flag or
+  # environment variable.
+  #
+  config.enable_analytics = proc { FeatureFlag.active?(:send_request_data_to_bigquery) }
```
Comment (JonF1) [inline config/initializers/dfe-analytics.rb:43]: Out of interest, why is this a proc? Wouldn't `FeatureFlag.active?(:send_request_data_to_bigquery)` also return true or false

Comment (tomas-stefano) [reply config/initializers/dfe-analytics.rb:43]: It is a proc because the block is saved to this enable_analytics config var so then to be called during request time and not when Rails is booting. So "save" the block here, call it later.
The method is defined here: https://github.com/DFE-Digital/dfe-analytics/blob/53efa7bfb759cdb1384133b11600eb041dfcbb03/lib/dfe/analytics.rb#L74

Comment (JonF1) [reply config/initializers/dfe-analytics.rb:43]: ah, thats makes sense, the proc is stored in var to be called later by that method `self.enabled?`. Thanks!

**Comment by misaka** [comment]
Just conscious that the specs part of this PR is basically one big delete. Is there any way to add meaningful tests? While working on ttapi I thought maybe we should include some kind of spec helper in this project that abstracts out the test that we can include in the implementing service. e.g. `have_sent_web_request_event`:

```ruby
    it "emits web request events using dfe-analytics" do
      expect {
        get "/test"
      }.to(have_sent_web_request_event)
    end
```

And then the app could decide whether to implement this in an example controller that inherits from `ApplicationController`, or wherever `DFE::Analytics::Requests` gets included, or to include it in tests for individual controllers.

It would also be a good idea to test that the DFE::Analytics is configured somehow. Just adding a spec to the app that tests whether `DFE::Analytics.config.bigquery_table_name` is set to something, for example, seems a bit redundant. Would it make sense for the generator to add a spec that tests that all the relevant `DFE::Analytics` configuration points are configured with non-nil values? This would trigger a test failure if the initializer isn't edited.

**Comment by duncanjbrown** [comment]
@misaka good points.

To the config one, `DfE::Analytics` will bail on send attempt if vital vars are missing: https://github.com/DFE-Digital/dfe-analytics/blob/main/lib/dfe/analytics.rb#L28

I agree we should have some sort of integration test in the app. I wonder if we should call `DfE::Analytics::Testing.fake!` and maybe that should count "requests"? Not sure if we should go all the way to the network boundary tho? (It does in the gem's own specs)

**Comment by JonF1** [comment]
Do we still need `Events::Event` class or can we delete it with its corresponding spec?

File: config/initializers/dfe-analytics.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  config.bigquery_dataset = ENV['BIG_QUERY_DATASET']
+
+  # Service account JSON key for the BigQuery API. See
+  # https://cloud.google.com/bigquery/docs/authentication/service-account-file
+  #
+  config.bigquery_api_json_key = ENV['BIG_QUERY_API_JSON_KEY']
+
+  # Passed directly to the retries: option on the BigQuery client
+  #
+  # config.bigquery_retries = 3
```
Comment (JonF1) [inline config/initializers/dfe-analytics.rb:33]: Since we had 3 retries in the deleted code, do we need this to be set for the new jobs? https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/workers/send_events_to_bigquery.rb#L5

Comment (duncanjbrown) [reply config/initializers/dfe-analytics.rb:33]: Good question. We don't actually have a "retries" value for the activejob job — maybe we should add a sensible default? wdyt?

**Comment by duncanjbrown** [comment]
> Do we still need `Events::Event` class or can we delete it with its corresponding spec?

Good catch! I'll delete :+1:

File: config/initializers/dfe-analytics.rb (hunk 2 above)
Comment (JonF1) [reply config/initializers/dfe-analytics.rb:33]: That makes sense, just reading on docs that sidekiq retries 25 times by default? but that seems excessive. I guess 3 or thereabouts would be more sensible

Comment (duncanjbrown) [reply config/initializers/dfe-analytics.rb:33]: yes that's a lot! I'll see if I can amend in gem

**Review by misaka — APPROVED** [review]
Looking good! Tested in review env and events appear to be coming through as before. We've also agreed to punt adding the discussed tests.
